File: backend/backend_idle/src/main/java/com/fullstack/entity/Order.java
// src/main/java/com/fullstack/entity/Order.java
package com.fullstack.entity;

import java.security.SecureRandom;
import java.time.LocalDateTime;

import org.hibernate.annotations.CreationTimestamp;
import org.hibernate.annotations.UpdateTimestamp;

import com.fullstack.model.enums.OrderStatus; // Added import

import jakarta.persistence.Column;
import jakarta.persistence.Entity;
import jakarta.persistence.EnumType;
import jakarta.persistence.Enumerated;
import jakarta.persistence.GeneratedValue;
import jakarta.persistence.GenerationType;
import jakarta.persistence.Id;
import jakarta.persistence.Index;
import jakarta.persistence.PrePersist;
import jakarta.persistence.Table;
import lombok.AllArgsConstructor;
import lombok.Builder;
import lombok.Getter;
import lombok.NoArgsConstructor;
import lombok.Setter;

@Entity
@Table(
    name = "orders",
    indexes = {
        // 주문번호 조회가 많으니 인덱스 + 유니크 권장
        @Index(name = "idx_orders_order_no", columnList = "order_no", unique = true)
    }
)
@Getter
@Setter
@NoArgsConstructor
@AllArgsConstructor
@Builder
public class Order {

    /* ===================== 기본키 ===================== */
    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    private Long id;

    /* ===================== 주문번호 ===================== */
    // 영문/숫자 랜덤, 접두사 ODR-, 날짜 안 붙음 (요구사항 B안)
    @Column(name = "order_no", length = 32, nullable = false, unique = true)
    private String orderNo;


    /** 금액/평균가 */
    @Column(name = "proposed_price")
    private Integer proposedPrice;        // 화주 제안가 (INTEGER)

    @Column(name = "driver_price")
    private Long driverPrice;             // 기사 확정가 (BIGINT)

    @Column(name = "avg_price")
    private Long avgPrice;                // 평균가 (BIGINT)

    /** 경로/거리 */
    private String departure;             // 출발지
    private String arrival;               // 도착지

    @Column(name = "distance")
    private Double distance;              // DOUBLE PRECISION (nullable 가능성 → Double)

    /** 예약/즉시 여부 */
    @Column(name = "reserved_date")
    private String reservedDate;          // VARCHAR

    @Column(name = "is_immediate")
    private Boolean isImmediate;          // BOOLEAN (nullable 가능성 → Boolean)

    /** 화물/차량/포장 */
    private String weight;                // VARCHAR
    private String vehicle;               // VARCHAR
    @Column(name = "cargo_type")
    private String cargoType;             // VARCHAR
    @Column(name = "cargo_size")
    private String cargoSize;             // VARCHAR

    @Column(name = "packing_option")
    private String packingOption;         // VARCHAR (콤마 문자열)

    /** 상태 */
	@Enumerated(EnumType.STRING) // Added annotation
	@Builder.Default
	private OrderStatus status = OrderStatus.CREATED; // Changed type and default value

    /* ===================== 경로/정보 ===================== */
    private String departure;           // 출발지
    private String arrival;             // 도착지
    private double distance;            // 거리(km 등)

    private String reservedDate;        // 예약 시간(문자열 보관 중이면 그대로 둠)

    // ⚠️ 서비스에서 dto.isImmediate()를 호출하므로 필드명은 isImmediate 그대로 유지
    private boolean isImmediate;        // 즉시 여부

    private String weight;              // 무게
    private String vehicle;             // 차량 종류
    private String cargoType;           // 화물 종류
    private String cargoSize;           // 화물 크기
    private String packingOption;       // 포장 옵션(콤마 문자열/요약)

    /* ===================== 배정 정보 ===================== */
    // 배정된 기사 ID (없으면 null)
    @Column(name = "assigned_driver_id")
    private Long assignedDriverId;

    @Column(name = "shipper_id")
    private String shipperId;             // 화주 ID


    /** 타임스탬프 */
    @CreationTimestamp
    @Column(updatable = false)
    private LocalDateTime createdAt;

    @UpdateTimestamp
    private LocalDateTime updatedAt;

    /* ===== 주문번호 자동 생성 ===== */
    @PrePersist
    public void ensureOrderNo() {
        // 상태 기본값 보정
        if (this.status == null) { // Check for null directly for enum
            this.status = OrderStatus.CREATED; // Set default to CREATED enum
        }
        // 주문번호 없으면 생성
        if (this.orderNo == null || this.orderNo.isBlank()) {
            this.orderNo = "ODR-" + randomCode(12);   // 총 길이 예: ODR-XXXXXXXXXXXX (16자)
        }
    }

    // 가독성 좋은 문자셋 (0/O, 1/I 제외)
    private static final String CHARS = "ABCDEFGHJKLMNPQRSTUVWXYZ23456789";
    private static final SecureRandom RND = new SecureRandom();

    private static String randomCode(int len) {
        StringBuilder sb = new StringBuilder(len);
        for (int i = 0; i < len; i++) {
            sb.append(CHARS.charAt(RND.nextInt(CHARS.length())));
        }
        return sb.toString();
    }

    // INSERT 직전(order_no가 비어 있으면) 자동 생성
    @PrePersist
    public void ensureOrderNo() {
        if (this.orderNo == null || this.orderNo.isBlank()) {
            // ODR- + 10자리(필요 시 길이 조절)
            this.orderNo = "ODR-" + randomCode(10);
        }
    }

}

File: application/lib/model/order.dart
class Order {
  final String id;
  final String orderNo;
  final String departure;
  final String arrival;
  final String status;
  final DateTime? createdAt;
  final DateTime? assignedAt;
  final DateTime? paidAt;
  final DateTime? departedAt;
  final DateTime? completedAt;
  final bool hasReview;
  final int? targetId;
  final int? driverPrice;

  // 🔽 추가된 화물 상세 필드들
  final double? distance; // 거리 (km)
  final String? cargoType; // 화물 종류 (박스, 팔레트 등)
  final String? cargoSize; // 화물 크기 (소형, 중형, 대형 등)
  final String? weight; // 무게 (문자열 저장: kg, ton 등)
  final String? vehicle; // 차량 종류 (1톤, 5톤, 트럭 등)
  final String? packingOption; // 포장 방식 (일반, 특수, 고가, 파손위험 등)
  final bool? isImmediate; // 즉시배송 여부 (true=즉시, false=예약)
  final String? reservedDate; // 예약일 (문자열 또는 DateTime 변환 필요)

  const Order({
    required this.id,
    required this.orderNo,
    required this.departure,
    required this.arrival,
    required this.status,
    this.createdAt,
    this.assignedAt,
    this.paidAt,
    this.departedAt,
    this.completedAt,
    this.hasReview = false,
    this.targetId,
    this.driverPrice,
    this.distance,
    this.cargoType,
    this.cargoSize,
    this.weight,
    this.vehicle,
    this.packingOption,
    this.isImmediate,
    this.reservedDate,
  });

  factory Order.fromJson(Map<String, dynamic> json) {
    print(json);
    return Order(
      id: json['id']?.toString() ?? '',
      orderNo: json['orderNo']?.toString() ?? '',
      departure: json['departure']?.toString() ?? '',
      arrival: json['arrival']?.toString() ?? '',
      status: json['status']?.toString() ?? 'NONE',
      createdAt: DateTime.tryParse(json['createdAt']?.toString() ?? ''),
      assignedAt: DateTime.tryParse(json['assignedAt']?.toString() ?? ''),
      paidAt: DateTime.tryParse(json['paidAt']?.toString() ?? ''),
      departedAt: DateTime.tryParse(json['departedAt']?.toString() ?? ''),
      completedAt: DateTime.tryParse(json['completedAt']?.toString() ?? ''),
      hasReview: json['hasReview'] == true,
      targetId: (json['assignedDriverId'] ?? json['targetId']) as int?,
      driverPrice: json['driverPrice'] as int?,
      distance: (json['distance'] is num)
          ? (json['distance'] as num).toDouble()
          : double.tryParse(json['distance']?.toString() ?? ''),
      cargoType: json['cargoType'] ?? json['cargo_type'],
      cargoSize: json['cargoSize'] ?? json['cargo_size'],
      weight: json['weight']?.toString(),
      vehicle: json['vehicle']?.toString(),
      packingOption: json['packingOption'] ?? json['packing_option'],
      isImmediate: json['isImmediate'] ?? json['is_immediate'],
      reservedDate: json['reservedDate'] ?? json['reserved_date'],
    );
  }

  Map<String, dynamic> toJson() => {
    'id': id,
    'orderNo': orderNo,
    'arrival': arrival,
    'status': status,
    'createdAt': createdAt?.toIso8601String(),
    'assignedAt': assignedAt?.toIso8601String(),
    'paidAt': paidAt?.toIso8601String(),
    'departedAt': departedAt?.toIso8601String(),
    'completedAt': completedAt?.toIso8601String(),
    'hasReview': hasReview,
    'targetId': targetId,
    'driverPrice': driverPrice,
    'distance': distance,
    'cargoType': cargoType,
    'cargoSize': cargoSize,
    'weight': weight,
    'vehicle': vehicle,
    'packingOption': packingOption,
    'isImmediate': isImmediate,
    'reservedDate': reservedDate,
  };

  Order copyWith({
    String? id,
    String? orderNo,
    String? departure,
    String? arrival,
    String? status,
    DateTime? createdAt,
    DateTime? assignedAt,
    DateTime? paidAt,
    DateTime? departedAt,
    DateTime? completedAt,
    bool? hasReview,
    int? targetId,
    int? driverPrice,
    double? distance,
    String? cargoType,
    String? cargoSize,
    String? weight,
    String? vehicle,
    String? packingOption,
    bool? isImmediate,
    String? reservedDate,
  }) {
    return Order(
      id: id ?? this.id,
      orderNo: orderNo ?? this.orderNo,
      departure: departure ?? this.departure,
      arrival: arrival ?? this.arrival,
      status: status ?? this.status,
      createdAt: createdAt ?? this.createdAt,
      assignedAt: assignedAt ?? this.assignedAt,
      paidAt: paidAt ?? this.paidAt,
      departedAt: departedAt ?? this.departedAt,
      completedAt: completedAt ?? this.completedAt,
      hasReview: hasReview ?? this.hasReview,
      targetId: targetId ?? this.targetId,
      driverPrice: driverPrice ?? this.driverPrice,
      distance: distance ?? this.distance,
      cargoType: cargoType ?? this.cargoType,
      cargoSize: cargoSize ?? this.cargoSize,
      weight: weight ?? this.weight,
      vehicle: vehicle ?? this.vehicle,
      packingOption: packingOption ?? this.packingOption,
      isImmediate: isImmediate ?? this.isImmediate,
      reservedDate: reservedDate ?? this.reservedDate,
    );
  }
}
